velox/expression/benchmarks/VariadicBenchmark.cpp

```diff
@@ -0,0 +1,208 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <folly/Benchmark.h>
+#include <folly/init/Init.h>
+
+#include "velox/functions/Macros.h"
+#include "velox/functions/Registerer.h"
+#include "velox/functions/lib/benchmarks/FunctionBenchmarkBase.h"
+#include "velox/functions/lib/string/StringImpl.h"
+
+// Benchmark a function that concats its string inputs.
+
+namespace facebook::velox::functions {
+
+template <typename T>
+struct StringConcatSimple {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& out,
+      const arg_type<Varchar>& str1,
+      const arg_type<Varchar>& str2,
+      const arg_type<Varchar>& str3,
+      const arg_type<Varchar>& str4,
+      const arg_type<Varchar>& str5) {
+    stringImpl::concatStatic(out, str1, str2, str3, str4, str5);
+
+    return true;
+  }
+};
+
+template <typename T>
+struct StringConcatVariadic {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& out,
+      const arg_type<Variadic<Varchar>>& strings) {
+    for (const auto& string : strings) {
+      const auto& value = string.value();
+      if (value.size() == 0) {
+        continue;
+      }
+      stringCore::applyAppendersRecursive(out, [&](out_type<Varchar>& out) {
+        auto writeOffset = out.size();
+        out.resize(out.size() + value.size());
+        std::memcpy(out.data() + writeOffset, value.data(), value.size());
+      });
+    }
+
+    return true;
+  }
+};
+
+void registerVectorFunction() {
+  VELOX_REGISTER_VECTOR_FUNCTION(udf_concat, "concat_vector");
+}
+
+void registerSimpleFunctions() {
+  registerFunction<
+      StringConcatSimple,
+      Varchar,
+      Varchar,
+      Varchar,
+      Varchar,
+      Varchar,
+      Varchar>({"concat_simple"});
+
+  registerFunction<StringConcatVariadic, Varchar, Variadic<Varchar>>(
+      {"concat_simple_variadic"});
+}
+
+namespace {
+
+class VariadicBenchmark : public functions::test::FunctionBenchmarkBase {
+ public:
+  VariadicBenchmark() : FunctionBenchmarkBase() {
+    registerVectorFunction();
+    registerSimpleFunctions();
+  }
+
+  RowVectorPtr makeData(vector_size_t numArgs) {
+    const vector_size_t size = 1'000;
+
+    std::vector<VectorPtr> args;
+
+    for (vector_size_t i = 0; i < numArgs; ++i) {
+      args.emplace_back(
+          vectorMaker_.flatVector<StringView>(size, [](vector_size_t row) {
+            return StringView(fmt::format("{}", row));
+          }));
+    }
+
+    return vectorMaker_.rowVector(args);
+  }
+
+  size_t run(const std::string& functionName, vector_size_t numArgs = 5) {
```

Comment:
note: no actions required.
I found that we can let folly decides on the number if iterations. using Benchmark(name, n)
but i have not tried it yet. I will try since it not sure how folly play with n.
see the benchmark here : https://github.com/facebookincubator/velox/pull/992